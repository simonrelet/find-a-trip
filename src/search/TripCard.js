import { format, isSameDay, parse } from 'date-fns';
import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components/macro';
import { CHILDREN_SPACING_SELECTOR, SPACING_UNIT } from '../styles';
import { UserInfo } from './UserInfo';

const TODAY = new Date();

const DATE_API_FORMAT = 'dd/MM/yyyy HH:mm:ss';
const TIME_FORMAT = 'HH:mm';
const DAY_FORMAT = 'EEEE do';

const Card = styled.div`
  padding: ${SPACING_UNIT * 2}em;
  background-color: white;
  box-shadow: 0px 2px 6px 0px rgba(0, 0, 0, 0.23);
  border-radius: ${SPACING_UNIT / 2}em;
  display: flex;
  flex-direction: column;
  align-items: stretch;

  ${CHILDREN_SPACING_SELECTOR} {
    margin-top: ${SPACING_UNIT * 2}em;
  }
`;

const CardContent = styled.div`
  display: flex;
  align-items: flex-start;
  justify-content: space-between;

  ${CHILDREN_SPACING_SELECTOR} {
    margin-left: ${SPACING_UNIT}em;
  }
`;

const CardFooter = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;

  ${CHILDREN_SPACING_SELECTOR} {
    margin-left: ${SPACING_UNIT}em;
  }
`;

const DepartureDate = styled.span`
  font-size: 1.25em;
`;

const TripInfo = styled.div`
  display: flex;
  flex-direction: column;

  ${CHILDREN_SPACING_SELECTOR} {
    margin-top: ${SPACING_UNIT}em;
  }
`;

const Price = styled.span`
  font-size: 1.25em;
  font-weight: bold;
`;

export function TripCard({ trip }) {
  const departureDate = parse(trip.departure_date, DATE_API_FORMAT, TODAY);
  let formattedDepartureTime = format(departureDate, TIME_FORMAT);

  if (!isSameDay(departureDate, TODAY)) {
    const day = format(departureDate, DAY_FORMAT);
    formattedDepartureTime = `${formattedDepartureTime} - ${day}`;
  }

  return (
    <Card>
      <CardContent>
        <TripInfo>
          <DepartureDate>{formattedDepartureTime}</DepartureDate>

          <span>
            {trip.departure_place.city_name} - {trip.arrival_place.city_name}
          </span>
        </TripInfo>

        <Price>{trip.price.string_value}</Price>
      </CardContent>

      <CardFooter>
        <UserInfo user={trip.user} />

        <a href={trip.links._front} target="_blank" rel="noopener noreferrer">
          See more on BlaBlaCar
        </a>
      </CardFooter>
    </Card>
  );
}

TripCard.propTypes = {
  trip: PropTypes.object.isRequired,
};
