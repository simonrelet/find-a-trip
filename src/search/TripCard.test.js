import { render } from '@testing-library/react';
import React from 'react';
import { TripCard } from './TripCard';

describe('TripCard', () => {
  it('should render the given trip', () => {
    const trip = {
      departure_date: '15/12/2019 05:00:00',
      departure_place: {
        city_name: 'Paris',
      },
      arrival_place: {
        city_name: 'Lyon',
      },
      price: {
        string_value: '22 €',
      },
      user: {
        links: {
          _front: 'https://url-to-user-profile.com',
        },
        has_picture: true,
        picture: 'https://url-to-user-picture.com',
        display_name: 'John Doe',
      },
      links: {
        _front: 'https://url-to-trip.com',
      },
    };

    const { container } = render(<TripCard trip={trip} />);
    expect(container).toMatchSnapshot();
  });
});
