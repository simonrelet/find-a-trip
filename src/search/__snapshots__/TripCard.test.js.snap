// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TripCard should render the given trip 1`] = `
<div>
  <div
    class="TripCard__Card-sc-1hl03w8-0 jSsuSi"
  >
    <div
      class="TripCard__CardContent-sc-1hl03w8-1 krkXMB"
    >
      <div
        class="TripCard__TripInfo-sc-1hl03w8-4 jDnUQd"
      >
        <span
          class="TripCard__DepartureDate-sc-1hl03w8-3 kCbXuN"
        >
          05:00 - Sunday 15th
        </span>
        <span>
          Paris
           - 
          Lyon
        </span>
      </div>
      <span
        class="TripCard__Price-sc-1hl03w8-5 bVihvi"
      >
        22 €
      </span>
    </div>
    <div
      class="TripCard__CardFooter-sc-1hl03w8-2 cgkZLn"
    >
      <a
        class="UserInfo__UserInfoWrapper-bs1fdn-0 crheuZ"
        href="https://url-to-user-profile.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="John Doe"
          class="UserInfo__Avatar-bs1fdn-1 kCcKVM"
          src="https://url-to-user-picture.com"
        />
        <span>
          John Doe
        </span>
      </a>
      <a
        href="https://url-to-trip.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        See more on BlaBlaCar
      </a>
    </div>
  </div>
</div>
`;
